src/sock/socket_session.cpp

```diff
@@ -149,6 +175,16 @@ void socket_session<T>::on_read(error_code ec, std::size_t)
         return fail(ec, "read");
     }
 
+    // Loop through thresholds, increment counter values of various thresholds and check whether counter exceeds threshold max value
+    for (auto &[key, threshold] : thresholds)
```

Comment:
No looping is needed here. Just get the threshold reference by vector index and increment the value. A better design would be to have a `.increment(<value>)` method in the `session_threshold` class. The `.increment()` method will perform exceeding checks and timer resets. (Basically you can move the logic in `threshold_validator` into increment() method)

Reply:
So you need to go through every threshold and increment e.g 
thresholds[util::THRESHOLDS::MAX_BYTES_PER_MINUTE].increment();

we have to do this for every threshold in thresholds vector. So whats the difference between looping and going one by one